app/controllers/concerns/two_factor_authentication_concern.rb

```diff
@@ -21,14 +37,30 @@ def valid_otp_attempt?(user)
   def authenticate_with_two_factor
     user = self.resource = find_user
 
-    if user_params[:otp_attempt].present? && session[:attempt_user_id]
-      authenticate_with_two_factor_attempt(user)
+    if user.webauthn_enabled? && user_params[:credential].present? && session[:attempt_user_id]
+      authenticate_with_two_factor_via_webauthn(user)
+    elsif user_params[:otp_attempt].present? && session[:attempt_user_id]
+      authenticate_with_two_factor_via_otp(user)
     elsif user.present? && user.external_or_valid_password?(user_params[:password])
       prompt_for_two_factor(user)
     end
   end
 
-  def authenticate_with_two_factor_attempt(user)
+  def authenticate_with_two_factor_via_webauthn(user)
+    webauthn_credential = WebAuthn::Credential.from_get(user_params[:credential])
+
+    if valid_webauthn_credential?(user, webauthn_credential)
+      session.delete(:attempt_user_id)
+      remember_me(user)
+      sign_in(user)
+      render json: { redirect_path: root_path }, status: :ok
+    else
+      flash.now[:alert] = t('webauthn_credentials.invalid_credential')
+      render json: {}, status: :unauthorized
```

Comment:
Same, the use of `flash.now` with a `json` reply seems suspicious

Reply:
`flash.now` will definitely not work with or without redirect.

Reply:
What we were trying to do here was to set the flash message and then reload the page using javascript so it renders properly, but, as you correctly point out this will not work using `flash.now`.

Anyway, as I mentioned in the other thread, we changed the way we displayed error messages so I will update this 🙂